File: LeetCode/Stack/JavaScript/validParentheses.js
// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.
// Every close bracket has a corresponding open bracket of the same type.

// Example 1:

// Input: s = "()"

// Output: true

// Example 2:

// Input: s = "()[]{}"

// Output: true

// Example 3:

// Input: s = "(]"

// Output: false

// Example 4:

// Input: s = "([])"

// Output: true

// Constraints:

// 1 <= s.length <= 104
// s consists of parentheses only '()[]{}'.

// TC: O(n), loopping through array once
// SC: O(n), we're maintaining a stack at worst could be size n.

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = [];
  const closeToOpen = { ")": "(", "]": "[", "}": "{" };
  // console.log(closeToOpen)

  for (let c of s) {
    // console.log(c)
    // First check if this is a closing bracket or parenthesis: ),},]
    if (c in closeToOpen) {
      // Since this is a closing bracket or parenthesis, we need to check the top or
      // our stack if it's an open bracket or parenthesis
      // If so, we pop our stack and move on. If not, then we don't have a valid parentheses
      if (stack.length !== 0 && stack.at(-1) === closeToOpen[c]) {
        stack.pop();
      } else {
        return false;
      }
    } else {
      // This is not a closing bracket of parenthesis, so we push it to our stack
      stack.push(c);
    }
  }
  // console.log(stack)
  // After we exit the for loop, we should have an empty stack. If not, then we did not have a valid parentheses
  return stack.length === 0 ? true : false;
};

File: LeetCode/Stack/python/validParentheses.py
from typing import Dict


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closeToOpen = {")": "(", "]": "[", "}": "{"}

        for c in s:
            # If c is a closing bracket, let's check the top of the 
            # stack if it's an openning bracket of similar kind and 
            # pop it from the stack. Else return false
            # If c is not a closing bracket, add it to the stack.
            if c in closeToOpen:
                if len(stack) > 0 and stack[-1] == closeToOpen[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        # If we popped everything from the stack, 
        # that means our input have valid parentheses
        return True if len(stack) == 0 else False
    
    @staticmethod
    def testSolution(input: Dict[str, str | bool]) -> None:
        print("Input: s: " + input["s"])
        print("Expected: " + str(input["expected"]))
        res = Solution().isValid(input["s"])
        print("Result: " + str(res))
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)
    

if __name__ == "__main__":
    input = {"s": "()", "expected": True}
    Solution.testSolution(input)

    input = {"s": "()[]{}", "expected": True}
    Solution.testSolution(input)

    input = {"s": "(]", "expected": False}
    Solution.testSolution(input)

    input = {"s": "([])", "expected": True}
    Solution.testSolution(input)

    input = {"s": "([)]", "expected": False}
    Solution.testSolution(input)